node/src/components/consensus/highway_core/state/tests.rs

```diff
@@ -188,30 +194,85 @@ fn add_unit() -> Result<(), AddUnitError<TestContext>> {
         endorsed: BTreeSet::new(),
     };
     let unit = SignedWireUnit::new(wunit.clone().into_hashed(), &BOB_SEC);
-    let maybe_err = state.add_unit(unit).err().map(unit_err);
+    let maybe_err = state.add_unit(unit, panorama.clone()).err().map(unit_err);
     assert_eq!(Some(UnitError::SequenceNumber), maybe_err);
+
     // Still not valid: This would be the third unit in the first round.
     wunit.seq_number = 2;
-    let unit = SignedWireUnit::new(wunit.into_hashed(), &BOB_SEC);
-    let maybe_err = state.add_unit(unit).err().map(unit_err);
+    let unit = SignedWireUnit::new(wunit.clone().into_hashed(), &BOB_SEC);
+    let maybe_err = state.add_unit(unit, panorama).err().map(unit_err);
     assert_eq!(Some(UnitError::ThreeUnitsInRound), maybe_err);
+    wunit.timestamp = 65.into();
+
+    // Your sequence number must match the panorama.
+    wunit.seq_number = 3;
+    let unit = SignedWireUnit::new(wunit.clone().into_hashed(), &BOB_SEC);
+    let maybe_err = state.pre_validate_unit(&unit, None).err();
+    assert_eq!(Some(UnitError::SequenceNumber), maybe_err);
+    wunit.seq_number = 2;
+
+    // If a hash-based panorama is included it must match the sequence number-based one.
+    let unit = SignedWireUnit::new(wunit.into_hashed(), &BOB_SEC);
+    let maybe_err = state
+        .add_unit(unit.clone(), panorama!(a0, b1, c0))
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PanoramaMismatch(ALICE)), maybe_err);
+    let maybe_err = state
+        .add_unit(unit.clone(), panorama!(N, b1, c0, N))
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PanoramaLength(4)), maybe_err);
+    let maybe_err = state
+        .add_unit(unit.clone(), panorama!(a0, b1, c0))
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PanoramaMismatch(ALICE)), maybe_err);
```

Comment:
Looks like it's duplicated with lines 216-220.

Reply:
Good catch! :+1: